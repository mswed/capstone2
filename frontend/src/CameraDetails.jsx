import { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { Container, Card, Row, Col } from 'react-bootstrap';
import GrumpyApi from './api';
import Loading from './Loading';
import FormatList from './FormatList.jsx';
import SimpleSearchFrom from './SimpleSearchForm.jsx';

const CameraDetails = () => {
  // Set up state
  const { cameraId } = useParams();
  const [cameraData, setCameraData] = useState(null);
  const [formatsData, setFormatsData] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  // Fetch camera data
  useEffect(() => {
    const getCameraDetails = async () => {
      try {
        const response = await GrumpyApi.getCameraDetails(cameraId);
        setCameraData(response);
        setFormatsData(response.formats);
      } catch (error) {
        console.error('Error fetching camera details', error);
      } finally {
        setIsLoading(false);
      }
    };
    getCameraDetails();
  }, [cameraId]);

  console.log('formatsData length', formatsData?.length);
  if (isLoading) {
    return <Loading />;
  }

  return (
    <Container>
      <Row className="g-4 shadow-lg">
        <Col md={4}>
          <Card.Img
            src={cameraData.image ? cameraData.image : '/media/camera_images/missing_image.png'}
            alt={`Logo for ${cameraData.model}`}
            className="h-100 py-3"
            style={{ objectFit: 'cover' }}
          />
        </Col>
        <Col md={4}>
          <Card.Title>{cameraData.model}</Card.Title>
          <Card.Body className="mt-3">
            <ul>
              <li>Make: {cameraData.make_name}</li>
              <li>Sensor Type: {cameraData.sensor_type}</li>
              <li>
                Max Filmback: {cameraData.max_filmback_width}mm x {cameraData.max_filmback_height}mm
              </li>
              <li>
                Max Resolution: {cameraData.max_image_width} x {cameraData.max_image_height}
              </li>
              <li>
                Frame Rate: {cameraData.min_frame_rate}fps - {cameraData.max_frame_rate}fps
              </li>
            </ul>
          </Card.Body>
        </Col>
        <Col md={4}>
          {cameraData.notes && (
            <div>
              <h6>Notes</h6>
              <p>{cameraData.notes}</p>
            </div>
          )}
        </Col>
      </Row>
      <Row className="mt-3">
        <Col>
          <Row>
            <SimpleSearchFrom
              fields={['image_format', 'image_aspect', 'format_name']}
              targetArray={formatsData}
              setTargetArray={setFormatsData}
              originalArray={cameraData.formats}
            />
          </Row>
          <Row>
            <FormatList formats={formatsData} />
          </Row>
        </Col>
      </Row>
    </Container>
  );
};

export default CameraDetails;
