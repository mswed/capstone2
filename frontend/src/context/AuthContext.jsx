import React, { useState, useEffect } from 'react';
import { jwtDecode } from 'jwt-decode';
import GrumpyApi from '../services/api';

// We create the context
const AuthContext = React.createContext();

/**
 * This is a logic only component that wraps our entire app and sets the
 * authentication context and its functions
 *
 * @param {Array}  children - Components to wrap
 * @returns {AuthContext.Provider} - with auth functions for easy access
 */

const AuthProvider = ({ children }) => {
  // We grab stuff from local storage if available
  const storedToken = localStorage.getItem('grumpy-token');
  const storedUser = localStorage.getItem('grumpy-user');

  // We set the token right away so we don't run in to authorization issues
  GrumpyApi.token = storedToken || null;

  // Initialize state
  const [token, setToken] = useState(storedToken);
  const [currentUser, setCurrentuser] = useState(storedUser);

  // Whenever the token changes (login/logout) we need to update
  // both the api module and localStorage
  useEffect(() => {
    if (token && token.trim !== '') {
      // We only set the token if it's not empty

      GrumpyApi.token = token;
      localStorage.setItem('grumpy-token', token || '');
      localStorage.setItem('grumpy-user', currentUser || '');
    } else {
      GrumpyApi.token = null;
      localStorage.removeItem('grumpy-token');
      localStorage.removeItem('grumpy-user');
    }
  }, [token]);

  // TODO:
  // // We also want to get the user favorites here
  // useEffect(() => {
  //   if (!token) return;
  //   async function getUserData() {
  //     const userData = await GrumpuyApi.getUser(currentUser);
  //     setFavorites(new Set(userData.favorites));
  //   }
  //   getUserData();
  // }, [token]);

  const login = async (username, password) => {
    try {
      const authToken = await GrumpyApi.login(username, password);
      setToken(authToken);
      const decoded = jwtDecode(authToken);
      setCurrentuser(decoded.username);
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const register = async (username, password, firstName, lastName, email) => {
    try {
      const token = await GrumpyApi.register(username, password, firstName, lastName, email);
      logout();
      return { success: true };
    } catch (error) {
      return { success: false, error };
    }
  };

  const logout = () => {
    // TODO: Do I need to logout of the backend here too?
    setToken('');
  };

  // TODO: We want to add favorites here
  // const addFavorite = async (jobId) => {
  //   const res = await GrumpuyApi.favoriteFormat(currentUser, jobId);
  //   if (res) {
  //     setFavorites((currentFavorites) => currentFavorites.add(res.favorites));
  //   }
  // };
  //

  // TODO: get items that have already been favorited
  // const alreadyFavorited = (formatId) => {
  //   return favFormats.has(formatId);
  // };

  // Create the value object
  const value = {
    token,
    currentUser,
    login,
    register,
    logout,
  };

  // Wrap everything in the context
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

export { AuthContext, AuthProvider };
