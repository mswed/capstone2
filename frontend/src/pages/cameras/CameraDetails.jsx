import { useState, useEffect } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import { Container, Card, Row, Col } from 'react-bootstrap';
import GrumpyApi from '../../services/api.js';
import Loading from '../../components/ui/Loading';
import FormatList from '../../features/formats/components/FormatList';
import LocalSearchForm from '../../components/forms/LocalSearchForm';
import ActionBar from '../../components/ui/ActionBar.jsx';
import ModalWindow from '../../components/ui/ModalWindow.jsx';
import CameraForm from '../../components/forms/CameraForm.jsx';
import FormatForm from '../../components/forms/FormatForm.jsx';
import ConfirmDialog from '../../components/ui/ConfirmDialog.jsx';
import useSources from '../../hooks/useSources.js';

const CameraDetails = () => {
  // Set up state
  const { cameraId } = useParams();
  const [cameraData, setCameraData] = useState(null);
  const [formatsData, setFormatsData] = useState([]);
  const [showEditCameraModal, setShowEditCameraModal] = useState(false);
  const [showNewFormatModal, setShowNewFormatModal] = useState(false);
  const [showConfirmDelete, setConfirmDelete] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  const { sources, setSources } = useSources();

  const navigate = useNavigate();

  const actionButtons = [
    {
      text: 'Delete Camera',
      variant: 'outline-danger',
      onClick: () => setConfirmDelete(true),
    },
    {
      text: 'Edit Camera',
      variant: 'outline-warning',
      onClick: () => setShowEditCameraModal(true),
    },
    {
      text: 'Add Format',
      variant: 'outline-success',
      onClick: () => setShowNewFormatModal(true),
    },
  ];

  const handleEditCamera = async (updatedCamera) => {
    try {
      // Cameras also take an image as one of their fields so we have to
      // use FormData instead of json
      const formData = new FormData();
      formData.append('make', cameraData.make);
      formData.append('model', updatedCamera.model);
      formData.append('sensor_type', updatedCamera.sensorType);
      formData.append('max_filmback_width', updatedCamera.maxFilmbackWidth);
      formData.append('max_filmback_height', updatedCamera.maxFilmbackHeight);
      formData.append('max_image_width', updatedCamera.maxImageWidth);
      formData.append('max_image_height', updatedCamera.maxImageHeight);
      formData.append('min_frame_rate', updatedCamera.minFrameRate);
      formData.append('max_frame_rate', updatedCamera.maxFrameRate);
      formData.append('notes', updatedCamera.notes);
      formData.append('discontinued', updatedCamera.discontinued);
      if (updatedCamera.image) {
        formData.append('image', updatedCamera.image);
      }

      const updatedCameraDetails = await GrumpyApi.updateCamera(cameraData.id, formData);
      setCameraData((prev) => ({ ...prev, ...updatedCameraDetails }));

      // Close the modal
      setShowEditCameraModal(false);
    } catch (error) {
      console.error('Failed to update camera:', error);
    }
  };

  const handleDeleteCamera = async () => {
    const response = await GrumpyApi.deleteCamera(cameraId);
    if (response.success) {
      navigate(-1);
    }
  };

  const handleAddFormat = async (newFormat) => {
    try {
      newFormat.camera = cameraId;
      await GrumpyApi.addFormat(newFormat);

      // Refresh the camera
      const response = await GrumpyApi.getCameraDetails(cameraId);
      setCameraData(response);
      setFormatsData(response.formats);

      // Close the modal
      setShowNewFormatModal(false);
    } catch (error) {
      console.error('Failed to create format:', error);
    }
  };

  const handleAddSource = async (sourceData) => {
    try {
      console.log('trying to add source');
      const newSource = await GrumpyApi.addSource(sourceData);
      setSources((prev) => [...prev, newSource]);

      // Retrun the new source so we can update the UI
      return newSource;
    } catch (error) {
      console.error('Failed to add source', error);

      throw error;
    }
  };

  // Fetch camera data
  useEffect(() => {
    const getCameraDetails = async () => {
      try {
        const response = await GrumpyApi.getCameraDetails(cameraId);
        setCameraData(response);
        setFormatsData(response.formats);
      } catch (error) {
        console.error('Error fetching camera details', error);
      } finally {
        setIsLoading(false);
      }
    };
    getCameraDetails();
  }, [cameraId]);

  if (isLoading) {
    return <Loading />;
  }

  return (
    <Container>
      <ConfirmDialog
        show={showConfirmDelete}
        title={`Delete camera "${cameraData.model}"`}
        message={'Are you sure?'}
        confirmText="Delete"
        onConfirm={handleDeleteCamera}
        onCancel={() => setConfirmDelete(false)}
      />
      <Card className="shadow-lg rounded-0 rounded-bottom">
        <Row className="g-0">
          <Col md={4}>
            <Card.Img
              src={cameraData.image ? cameraData.image : '/media/camera_images/missing_image.png'}
              alt={`Logo for ${cameraData.model}`}
              className="p-3"
              style={{
                objectFit: 'contain',
                maxHeight: '250px',
                maxWidth: '100%',
              }}
            />
          </Col>
          <Col md={5}>
            <Card.Body className="text-start">
              <Card.Title>{cameraData.model}</Card.Title>
              <dl className="row">
                <dt className="col-sm-4">Make:</dt>
                <dd className="col-sm-8">{cameraData.make_name}</dd>

                <dt className="col-sm-4">Sensor Type:</dt>
                <dd className="col-sm-8">{cameraData.sensor_type}</dd>

                <dt className="col-sm-4">Max Filmback:</dt>
                <dd className="col-sm-8">
                  {cameraData.max_filmback_width}mm x {cameraData.max_filmback_height}mm
                </dd>

                <dt className="col-sm-4">Max Resolution:</dt>
                <dd className="col-sm-8">
                  {cameraData.max_image_width} x {cameraData.max_image_height}
                </dd>

                <dt className="col-sm-4">Frame Rate:</dt>
                <dd className="col-sm-8">
                  {cameraData.min_frame_rate}fps - {cameraData.max_frame_rate}
                  fps
                </dd>
              </dl>
            </Card.Body>
          </Col>
          <Col md={3} className="text-start p-3">
            {cameraData.notes && (
              <div>
                <h6>Notes</h6>
                <p>{cameraData.notes}</p>
              </div>
            )}
          </Col>
        </Row>
      </Card>
      <ActionBar buttons={actionButtons} className="mt-3" />
      <ModalWindow
        show={showEditCameraModal}
        onHide={() => setShowEditCameraModal(false)}
        title={`Edit ${cameraData.make_name} ${cameraData.model}`}
        form={<CameraForm onSubmit={handleEditCamera} camData={cameraData} />}
      />
      <ModalWindow
        show={showNewFormatModal}
        onHide={() => setShowNewFormatModal(false)}
        title={`Add format to ${cameraData.make_name} ${cameraData.model}`}
        form={<FormatForm onSubmit={handleAddFormat} camData={cameraData} sources={sources} onSourceAdded={handleAddSource} />}
      />
      {cameraData.formats.length > 0 && (
        <Row className="mt-3">
          <Col>
            <Row>
              <LocalSearchForm
                fields={['image_format', 'image_aspect', 'format_name']}
                targetArray={formatsData}
                setTargetArray={setFormatsData}
                originalArray={cameraData.formats}
              />
            </Row>
            <Row>
              <FormatList formats={formatsData} />
            </Row>
          </Col>
        </Row>
      )}
    </Container>
  );
};

export default CameraDetails;
