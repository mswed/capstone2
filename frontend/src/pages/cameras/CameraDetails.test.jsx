import { vi, it, expect, beforeEach } from 'vitest';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import CameraDetails from './CameraDetails.jsx';
import { AuthContext } from '../../context/AuthContext.jsx';
import { MessagesProvider } from '../../context/MessageContext.jsx';
import GrumpyApi from '../../services/api.js';

// Mock API calls
vi.mock('../../services/api', () => ({
  default: {
    getCameraDetails: vi.fn(() =>
      Promise.resolve({
        id: 1,
        makeId: 1,
        makeName: 'ARRI',
        model: 'Alexa 35',
        sensorType: 'Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array',
        sensorSize: 'Super 35',
        maxFilmbackWidth: 27.99,
        maxFilmbackHeight: 19.22,
        maxImageWidth: 4608,
        maxImageHeight: 3164,
        minFrameRate: 0.75,
        maxFrameRate: 120,
        image: 'arri_alexa_35.png',
        formats: [
          {
            id: 1,
            camera: 1,
            cameraModel: 'Alexa 35',
            makeName: 'ARRI',
            imageFormat: '4.6K',
            imageAspect: '3:2',
            formatName: 'Open Gate',
            sensorWidth: 28.0,
            sensorHeight: 19.2,
            imageWidth: 4608,
            imageHeight: 3164,
            pixelAspect: 1.0,
            isAnamorphic: false,
            isDesqueezed: false,
            anamorphicSqueeze: null,
            filmbackWidth3de: 28.0,
            filmbackHeight3de: 19.2,
            isDownsampled: false,
            isUpscaed: false,
            codec: 'ARRIRAW',
            rawReordingAvailable: true,
            source: 1,
            notes: 'sample notes',
            makeNotes:
              '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
            trackingWorkflow: 'Do this then that',
          },
          {
            id: 2,
            camera: 1,
            imageFormat: '4K',
            imageAspect: '16:9',
            sensorWidth: 28.0,
            sensorHeight: 15.7,
            imageWidth: 4096,
            imageHeight: 2304,
            pixelAspect: 1.0,
            isAnamorphic: false,
            isDesqueezed: false,
            anamorphicSqueeze: null,
            filmbackWidth3de: 28.0,
            filmbackHeight3de: 19.2,
            isDownsampled: true,
            isUpscaed: false,
            codec: 'ProRes',
            rawReordingAvailable: true,
            source: 1,
            notes: 'Downsampled from 4.6K 16:9',
            makeNotes:
              '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
            trackingWorkflow: 'Do that then this',
          },
        ],
      })
    ),
    addFormat: vi.fn(() =>
      Promise.resolve({
        success: 'Created format 7',
        formatId: 7,
      })
    ),
    updateCamera: vi.fn(() =>
      Promise.resolve({
        success: 'Partialy updated camera',
        camera: {
          id: 1,
          makeId: 1,
          makeName: 'ARRI',
          model: 'Updated Alexa 35',
          sensorType: 'Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array',
          sensorSize: 'Super 35',
          maxFilmbackWidth: 27.99,
          maxFilmbackHeight: 19.22,
          maxImageWidth: 4608,
          maxImageHeight: 3164,
          minFrameRate: 0.75,
          maxFrameRate: 120,
          image: 'arri_alexa_35.png',
          formats: [
            {
              id: 1,
              camera: 1,
              cameraModel: 'Alexa 35',
              makeName: 'ARRI',
              imageFormat: '4.6K',
              imageAspect: '3:2',
              formatName: 'Open Gate',
              sensorWidth: 28.0,
              sensorHeight: 19.2,
              imageWidth: 4608,
              imageHeight: 3164,
              pixelAspect: 1.0,
              isAnamorphic: false,
              isDesqueezed: false,
              anamorphicSqueeze: null,
              filmbackWidth3de: 28.0,
              filmbackHeight3de: 19.2,
              isDownsampled: false,
              isUpscaed: false,
              codec: 'ARRIRAW',
              rawReordingAvailable: true,
              source: 1,
              notes: 'sample notes',
              makeNotes:
                '4.6K 3:2 Open Gate provides maximum image quality, resolution, and flexibility in post for many spherical and anamorphic lenses in an image area slightly larger than traditional Super 35 film specifications.',
              trackingWorkflow: 'Do this then that',
            },
          ],
        },
      })
    ),
    deleteCamera: vi.fn(() =>
      Promise.resolve({
        success: 'Camera deleted',
      })
    ),
    getSources: vi.fn(() =>
      Promise.resolve([
        {
          id: 1,
          name: 'Sample Source',
          url: 'http://www.samplesource.com',
          fileName: 'sample.pdf',
          note: 'a fake source',
        },
      ])
    ),
  },
}));

describe('CameraDetails component', () => {
  /**
   * Render the componenet with authentication context
   *
   * @returns {Component} - Makes component wrapped in MemoryRouter and AuthContext
   */

  const renderWithAuth = (authValue = { token: 'fake-token', currentUser: 'testUser' }) => {
    return render(
      <MemoryRouter initialEntries={['/cameras/1']}>
        <MessagesProvider>
          <AuthContext.Provider value={authValue}>
            <Routes>
              <Route path="/cameras/:cameraId" element={<CameraDetails />} />
            </Routes>
          </AuthContext.Provider>
        </MessagesProvider>
      </MemoryRouter>
    );
  };

  beforeEach(() => {
    // Clear mocks
    vi.clearAllMocks();
  });

  it('renders without crashing', async () => {
    renderWithAuth();
    await screen.findByText('Alexa 35');
  });

  it('matches snapshot', async () => {
    const { asFragment } = renderWithAuth();

    await screen.findByText('Alexa 35');
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays the correct info', async () => {
    renderWithAuth();

    await waitFor(() => {
      expect(screen.getByText('ARRI')).toBeInTheDocument();
      expect(screen.getByText('Alexa 35')).toBeInTheDocument();
      expect(screen.getByText('Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array')).toBeInTheDocument();
      expect(screen.getByText('4.6K 3:2 Open Gate')).toBeInTheDocument();
    });
  });

  it('shows action bar for admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'admminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');
    expect(screen.getByText('Add Format')).toBeInTheDocument();
  });

  it('hides action bar for regular users', async () => {
    renderWithAuth({ token: 'fake-token', currentUser: 'admminUser' });
    await screen.findByText('ARRI');
    expect(screen.queryByText('Add Format')).not.toBeInTheDocument();
  });

  it('can add a new format when admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'adminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');

    // Open modal
    fireEvent.click(screen.getByText('Add Format'));
    expect(screen.getByRole('textbox', { name: /Image Format/i })).toBeInTheDocument();
    expect(screen.getByRole('textbox', { name: /Image Aspect/i })).toBeInTheDocument();

    // Fill out the required fields
    const sourceInput = screen.getByLabelText(/Source/i);
    const imageFormatInput = screen.getByLabelText(/Image Format/i);
    const imageAspecInput = screen.getByLabelText(/Image Aspect/i);
    const imageWidthInput = screen.getByLabelText(/Image Width/i);
    const imageHeightInput = screen.getByLabelText(/Image Height/i);
    const sensorWidthInput = screen.getByLabelText(/Sensor Width/i);
    const sensorHeightInput = screen.getByLabelText(/Sensor Height/i);

    fireEvent.change(sourceInput, { target: { value: 'Fake Source' } });
    fireEvent.change(imageFormatInput, {
      target: {
        value: '5K',
      },
    });
    fireEvent.change(imageAspecInput, {
      target: {
        value: '3:2',
      },
    });
    fireEvent.change(imageWidthInput, {
      target: {
        value: 4608,
      },
    });
    fireEvent.change(imageHeightInput, {
      target: { value: 3164 },
    });
    fireEvent.change(sensorWidthInput, {
      target: {
        value: 28.0,
      },
    });
    fireEvent.change(sensorHeightInput, {
      target: {
        value: 19.2,
      },
    });

    // Submit the form
    fireEvent.click(screen.getByText('Add'));

    // Modal should close after successful submission
    await waitFor(() => {
      expect(screen.queryByText('Add format to Arri Alexa 35')).not.toBeInTheDocument();
    });
  });

  it('can edit the camera when admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'adminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');

    // Open modal
    fireEvent.click(screen.getByText('Edit Camera'));
    expect(screen.getByText('Edit ARRI Alexa 35')).toBeInTheDocument();

    // Fill out the required fields
    const nameInput = screen.getByLabelText(/model/i);

    fireEvent.change(nameInput, { target: { value: 'Updated Camera' } });

    // Submit the form
    fireEvent.click(screen.getByText('Update'));

    // Modal should close after successful submission
    await waitFor(() => {
      expect(screen.queryByText('Edit ARRI Alexa 35')).not.toBeInTheDocument();
    });
  });

  it('can delete the camera when admin', async () => {
    renderWithAuth({
      token: 'fake-token',
      currentUser: 'adminUser',
      isAdmin: true,
    });
    await screen.findByText('ARRI');

    // Open modal
    fireEvent.click(screen.getByText('Delete Camera'));
    expect(screen.getByText('Delete camera "Alexa 35"')).toBeInTheDocument();

    // Click the button
    fireEvent.click(screen.getByText('Delete'));

    // Modal should close after successful submission
    await waitFor(() => {
      expect(GrumpyApi.deleteCamera).toHaveBeenCalledWith('1');
    });
  });
});
