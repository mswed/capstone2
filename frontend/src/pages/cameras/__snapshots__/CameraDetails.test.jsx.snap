// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CameraDetails component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="shadow-lg rounded-0 rounded-bottom card"
    >
      <div
        class="g-0 row"
      >
        <div
          class="col-md-4"
        >
          <img
            alt="Logo for Alexa 35"
            class="card-img p-3"
            src="arri_alexa_35.png"
            style="object-fit: contain; max-height: 250px; max-width: 100%;"
          />
        </div>
        <div
          class="col-md-5"
        >
          <div
            class="text-start card-body"
          >
            <div
              class="card-title h5"
            >
              Alexa 35
            </div>
            <dl
              class="row"
            >
              <dt
                class="col-sm-4"
              >
                Make:
              </dt>
              <dd
                class="col-sm-8"
              >
                ARRI
              </dd>
              <dt
                class="col-sm-4"
              >
                Sensor Type:
              </dt>
              <dd
                class="col-sm-8"
              >
                Super 35 format ARRI ALEV 4 CMOS sensor with Bayer pattern color filter array
              </dd>
              <dt
                class="col-sm-4"
              >
                Max Filmback:
              </dt>
              <dd
                class="col-sm-8"
              >
                27.99mm x 19.22mm
              </dd>
              <dt
                class="col-sm-4"
              >
                Max Resolution:
              </dt>
              <dd
                class="col-sm-8"
              >
                4608 x 3164
              </dd>
              <dt
                class="col-sm-4"
              >
                Frame Rate:
              </dt>
              <dd
                class="col-sm-8"
              >
                0.75fps - 120fps
              </dd>
            </dl>
          </div>
        </div>
        <div
          class="text-start p-3 col-md-3"
        />
      </div>
    </div>
    <div
      class="mt-3 row"
    >
      <div
        class="col"
      >
        <div
          class="row"
        >
          <form
            class=""
          >
            <div
              class="my-3 input-group"
            >
              <input
                class="form-control"
                placeholder="Search formats"
                value=""
              />
            </div>
          </form>
        </div>
        <div
          class="row"
        >
          <div
            class="table-responsive"
          >
            <table
              class="small table table-sm table-striped table-hover"
            >
              <thead>
                <tr>
                  <th>
                    Format
                  </th>
                  <th>
                    Resolution
                  </th>
                  <th>
                    Filmback
                  </th>
                  <th>
                    PAR
                  </th>
                  <th>
                    Anamorphic?
                  </th>
                  <th>
                    Desqueezed?
                  </th>
                  <th>
                    Action
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    4.6K 3:2 Open Gate
                  </td>
                  <td>
                    4608 x 3164
                  </td>
                  <td>
                    28mm x 19.2mm
                  </td>
                  <td>
                    1
                  </td>
                  <td
                    class="text-center"
                  >
                    <span
                      class="text-danger"
                    >
                      ✗
                    </span>
                  </td>
                  <td
                    class="text-center"
                  >
                    <span
                      class="text-danger"
                    >
                      ✗
                    </span>
                  </td>
                  <td>
                    <a
                      class="btn btn-outline-primary btn-sm"
                      data-discover="true"
                      href="/formats/1"
                    >
                      View
                    </a>
                  </td>
                </tr>
                <tr>
                  <td>
                    4K 16:9 
                  </td>
                  <td>
                    4096 x 2304
                  </td>
                  <td>
                    28mm x 15.7mm
                  </td>
                  <td>
                    1
                  </td>
                  <td
                    class="text-center"
                  >
                    <span
                      class="text-danger"
                    >
                      ✗
                    </span>
                  </td>
                  <td
                    class="text-center"
                  >
                    <span
                      class="text-danger"
                    >
                      ✗
                    </span>
                  </td>
                  <td>
                    <a
                      class="btn btn-outline-primary btn-sm"
                      data-discover="true"
                      href="/formats/2"
                    >
                      View
                    </a>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
